core/src/main/scala/kafka/utils/SelectorUtils.scala

```diff
@@ -0,0 +1,62 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kafka.utils
+
+import java.util.concurrent.TimeUnit
+
+import org.apache.kafka.common.network.Selector
+import org.apache.kafka.common.utils.{Time => JTime}
+
+object SelectorUtils {
+
+  /**
+   * Invokes `selector.poll` until `predicate` returns `true` or the timeout expires. It returns `true` if the call
+   * completes normally or `false` if the timeout expires.
+   */
+  def pollUntil(selector: Selector, timeout: Long)(predicate: => Boolean)(implicit time: JTime): Boolean =
+    pollUntilFound(selector, timeout) {
+      if (predicate) Some(true)
+      else None
+    }.fold(false)(_ => true)
+
+  /**
+   * Invokes `selector.poll` until `find` returns `Some` or the timeout expires. It returns the result of `find` if
+   * the call completes normally or `None` if the timeout expires.
+   */
+  def pollUntilFound[T](selector: Selector, timeout: Long)(find: => Option[T])(implicit time: JTime): Option[T] = {
```

Comment:
I think these utils are very generic, but also make the code a bit harder to read.

It looks like the "find" we are waiting for is basically always a reply from a specific broker? In this case, maybe make the generic method private and provide a more specific wrapper?